### PR TITLE
Core interceptor refactor.

### PR DESCRIPTION
**Pull request by dlorenc**
# Changes

This starts a refactor to support "core" event interceptors. This moves
the existing "ObjectRef" style interceptor to a new package called "webhook",
and embeds it as a "WebhookInterceptor" on the EventBinding type.

Other interceptors (Github, Gitlab, etc.) will be added alongside this. This
PR also extracts the beginnings of the "Interceptor" interface that these other
interceptors will implement.

Ref #207 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

**Comment by vtereso** [comment]
An EL trigger can specify an event interceptor, which is usually a webhook event, but does not have to be. Do you think it makes sense to use a different name?

File: pkg/apis/triggers/v1alpha1/event_listener_types.go [hunk 1]
```diff
@@ -77,6 +77,12 @@ type EventInterceptor struct {
 	// interceptor request headers. This allows the interceptor to make
 	// decisions specific to an EventListenerTrigger.
 	Header []pipelinev1.Param `json:"header,omitempty"`
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_types.go]: Does the Header field only apply to Webhook interceptors? If so, probably should be within the `WebhookInterceptor` struct?

**Comment by vtereso** [comment]
I had some thoughts more thoughts regarding the interceptor schema. IIUC, the core interceptors for GitHub, GitLab, etc. would be simple validator interceptors for their respective webhooks. In this way, maybe we should explicitly separate the interceptors? For example, 
```
interceptors:
    validation:
    mutator:
```
where validators would likely only be the core interceptors since we would want to provide easy support out of the box? I don't think this is necessary and we really wouldn't have a way to validate an interceptor is a validation/mutating interceptor, but just thought I'd mention.

**Comment by dlorenc** [comment]
> I had some thoughts more thoughts regarding the interceptor schema. IIUC, the core interceptors for GitHub, GitLab, etc. would be simple validator interceptors for their respective webhooks. In this way, maybe we should explicitly separate the interceptors? For example,

Great point! I think actually @wlynch had some ideas in mind where the github interceptor would also modify the incoming request. Use cases included adding a bit on whether or not the sender of the PR is a collaborator on the repo, which is a separate set of github API calls.

File: pkg/interceptors/interceptors.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"net/http"
+
+	triggersv1 "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+)
+
+// InterceptorInterface is the interface that all interceptors implement.
+type InterceptorInterface interface {
+	ExecuteTrigger([]byte, *http.Request, triggersv1.EventListenerTrigger, string) ([]byte, error)
```
[the remark this answers is not in the document]
Comment (dlorenc) [reply pkg/interceptors/interceptors.go]: D'oh, added names - hopefully they help.

File: pkg/apis/triggers/v1alpha1/event_listener_types.go (hunk 1 above)
Comment (dlorenc) [reply pkg/apis/triggers/v1alpha1/event_listener_types.go]: Yeah I think so. I misunderstood what it did earlier.

File: pkg/interceptors/interceptors.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+package interceptors
+
+import (
+	"net/http"
+
+	triggersv1 "github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+)
+
+// InterceptorInterface is the interface that all interceptors implement.
+type InterceptorInterface interface {
```
Comment (dibyom) [inline pkg/interceptors/interceptors.go]: Could this just be Interface instead of InterceptorInterface?

Comment (bigkevmcd) [reply pkg/interceptors/interceptors.go]: Or how about just `Interceptor`?

Comment (dlorenc) [reply pkg/interceptors/interceptors.go]: Done!

**Pull request merged**